### PR TITLE
Free variables: fix #909 and #911 by ignoring casts in EquivExprs

### LINKED ISSUE #909: Add lvalue DeclRefExprs to variable set for diagnosing free variables
Currently in #903, getting the set of all variables involved in an expression constructs an LValueToRValue ImplicitCastExpr for each variable to add to the set. This is done so that the expressions in the set are rvalues and can later be compared with other rvalue expressions in `CheckingState.EquivExprs`. We want to avoid constructing these ImplicitCastExprs.

`CollectVariableSetHelper` should collect a set of DeclRefExprs. To check whether there is a set in `CheckingState.EquivExprs` that contains an lvalue variable `v`, search each expression `e` in each set `F` in `CheckingState.EquivExprs` to see if `e` is an LValueToRValue cast of `v`. The `IsRValueCastOfVariable(e, v)` method can be used to check this.

### PR DESCRIPTION
**Pull request by dopelsunce**
On checking free variables, create a temporary `EquivVars` that only holds DecRefExpr and IntegerLiteral from the `EquivExprs`, with casts ignored. The variables in `EquivVars` are then used to compare with the variables (DecRefExpr) collected from bounds expressions.

IntegerLiterals from `EquivExprs` are also kept in `EquivVars` so we can check whether a variable equals to a constant.

### CODE REVIEW

File: clang/lib/Sema/SemaBounds.cpp [hunk 1]
```diff
@@ -1492,10 +1487,24 @@ namespace {
...
 
-        if (AddFreeVariables(S, Vars1, Vars2, EquivExprs, Pos1, FreeVars))
+        // EquivVars holds sets of DeclRefExpr and IntegerLiteral filtered from
+        // EquivExprs.
+        EquivExprSets EquivVars;
+        for (auto ExprSet : *EquivExprs) {
+          EqualExprTy Vars;
+          auto It = ExprSet.begin();
+          for (; It != ExprSet.end(); It++) {
+            *It = (*It)->IgnoreCasts();
```
Comment (kkjeer) [inline clang/lib/Sema/SemaBounds.cpp:1497]: This also needs to ignore parentheses: 

```
Expr *E = (*It)->IgnoreParenCasts();
if (isa<IntegerLiteral>(E) || isa<DeclRefExpr>(E))
  Vars.push_back(E);
``` 

Otherwise, the following test case fails:

```
int a1 checked[5];
short int t20 : byte_count(5 * sizeof(int)) = ((((short int)(a1))));
```

File: clang/lib/AST/CanonBounds.cpp [hunk 2]
```diff
@@ -310,7 +310,7 @@ Result Lexicographic::CompareExpr(const Expr *Arg1, const Expr *Arg2) {
 #include "clang/AST/StmtNodes.inc"
        llvm_unreachable("cannot compare a statement");  
      case Expr::PredefinedExprClass: Cmp = Compare<PredefinedExpr>(E1, E2); break;
-     case Expr::DeclRefExprClass: return Compare<DeclRefExpr>(E1, E2);
+     case Expr::DeclRefExprClass: Cmp = Compare<DeclRefExpr>(E1, E2); break;
```
Comment (kkjeer) [inline clang/lib/AST/CanonBounds.cpp:313]: Why this change?

**Review by kkjeer — COMMENTED** [review]
Thanks for working on this! I have a comment about removing parentheses.

File: clang/lib/AST/CanonBounds.cpp (hunk 2 above)
Comment (dopelsunce) [reply clang/lib/AST/CanonBounds.cpp:313]: Previously the code just returns here. This adds a new path so that if `Cmp` is not Equal, we proceed to use `CheckEquivExprs` to check if E1 and E2 produce the same value using EquivExprSets. (https://github.com/microsoft/checkedc-clang/pull/919/files#diff-9ff223384e8cae5a65b76ee59e13437eR376)

Comment (kkjeer) [reply clang/lib/AST/CanonBounds.cpp:313]: Is there a test case that fails if this change is not made?

Comment (dopelsunce) [reply clang/lib/AST/CanonBounds.cpp:313]:
```
void f() {
  int i1, i2;
  i1 = i2;
  _Array_ptr<int> p1 : count(i1 + 1) = 0;
  _Array_ptr<int> p2 : count(i2) = p1;
}
```

Without this change, `i1` and `i2` will be incorrectly reported as free variables in the code above, because`Compare<DeclRefExpr>` does not use the information from EquivExprSet to check if `i1` and `i2` can produce the same value. Once  `Compare<DeclRefExpr>` returns false, `CheckEquivExprs` also returns without using the information from the sets of equivalent expressions.

Previously in #903, we compare two ImplicitCastExprs rather than two DeclRefExprs, so this change was not needed. Now we directly compare DeclRefExprs collected from bounds expression, which is a different case for CheckEquivExprs.

Comment (kkjeer) [reply clang/lib/AST/CanonBounds.cpp:313]: Thanks for clarifying!

**Review by kkjeer — APPROVED** [review]
Thanks for fixing this!
